Add documentations for new MSTest rules (#40076)
dotnet/docs: docs/core/testing/mstest-analyzers/mstest0019.md

helpviewer_keywords:
- PreferTestInitializeOverConstructorAnalyzer
- MSTEST0019
author: Evangelink
ms.author: amauryleve
---
# MSTEST0019: Prefer TestInitialize methods over constructors

| Property                            | Value                                           |
|-------------------------------------|-------------------------------------------------|
| **Rule ID**                         | MSTEST0019                                      |
| **Title**                           | Prefer TestInitialize methods over constructors |
| **Category**                        | Design                                          |
| **Fix is breaking or non-breaking** | Non-breaking                                    |
| **Enabled by default**              | No                                              |
| **Default severity**                | Info                                            |
| **Introduced in version**           | 3.4.0                                           |

## Cause

This rule raises a diagnostic when there is a parameterless explicit constructor declared on a test class (class marked with `[TestClass]`).

## Rule description

> Some developers prefer to always use `[TestInitialize]` method instead of constructors for their test setup as it provides a symmetrical initialization to doing async/await initialization where you do require `[TestInitialize]` methods.

Reviewer: ```suggestion
Some developers prefer to always use `[TestInitialize]` method instead of constructors for their test setup as it provides a unified approach to initialization when doing async/await initialization that requires `[TestInitialize]` methods.
```

Reviewer: Or maybe
```suggestion
Use this rule to enforce using `[TestInitialize]` for both synchronous and asynchronous test initialization. Asynchronous 
 (async/await) test intialization requires the use of `[TestInitialize]` methods, because the resulting `Task` needs to be awaited.
```

I won't blame you if you won't merge either, I am not super happy with either of those suggestions.

helpviewer_keywords:
- PreferTestInitializeOverConstructorAnalyzer
- MSTEST0019
author: Evangelink
ms.author: amauryleve
---
# MSTEST0019: Prefer TestInitialize methods over constructors

| Property                            | Value                                           |
|-------------------------------------|-------------------------------------------------|
| **Rule ID**                         | MSTEST0019                                      |
| **Title**                           | Prefer TestInitialize methods over constructors |
| **Category**                        | Design                                          |
| **Fix is breaking or non-breaking** | Non-breaking                                    |
| **Enabled by default**              | No                                              |
| **Default severity**                | Info                                            |
| **Introduced in version**           | 3.4.0                                           |

## Cause

This rule raises a diagnostic when there is a parameterless explicit constructor declared on a test class (class marked with `[TestClass]`).

## Rule description

> Use this rule to enforce using `[TestInitialize]` for both synchronous and asynchronous test initialization. Asynchronous 
>  (async/await) test intialization requires the use of `[TestInitialize]` methods, because the resulting `Task` needs to be awaited.

Author: ```suggestion
Use this rule to enforce using `[TestInitialize]` for both synchronous and asynchronous test initialization. Asynchronous (async/await) test intialization requires the use of `[TestInitialize]` methods, because the resulting `Task` needs to be awaited.
```
